### DISCUSSION TITLE
Reduce count of GraphicsPipelineConfigurator

### DISCUSSION THREAD

**Discussion by avilkinnick** [vsg-dev/VulkanSceneGraph#1213] [Q&A]
Hello, now i am loading may models like pagedLod and for every of them I am creating new GraphicsPipelineConfigurator like this:
```
  auto pipeline = vsg::GraphicsPipelineConfigurator::create(options->shaderSets.at("phong"));

  vsg::DataList vertexArrays;
  pipeline->assignArray(vertexArrays, "vsg_Vertex", VK_VERTEX_INPUT_RATE_VERTEX, model_data->vertices);
  pipeline->assignArray(vertexArrays, "vsg_Normal", VK_VERTEX_INPUT_RATE_VERTEX, model_data->normals);
  pipeline->assignArray(vertexArrays, "vsg_TexCoord0", VK_VERTEX_INPUT_RATE_VERTEX, model_data->tex_coords);
  pipeline->assignArray(vertexArrays, "vsg_Color", VK_VERTEX_INPUT_RATE_VERTEX, model_data->colors);

  auto drawCommands = vsg::Commands::create();
  drawCommands->addChild(vsg::BindVertexBuffers::create(pipeline->baseAttributeBinding, vertexArrays));
  drawCommands->addChild(vsg::BindIndexBuffer::create(model_data->indices));
  drawCommands->addChild(vsg::DrawIndexed::create(model_data->indices->size(), 1, 0, 0, 0));

  if (texture_data)
  {
      static auto sampler = vsg::Sampler::create();
      pipeline->assignTexture("diffuseMap", texture_data, sampler);
  }

  pipeline->init();

  auto state_group = vsg::StateGroup::create();
  pipeline->copyTo(state_group);
  state_group->addChild(drawCommands);
```
```
    auto phong = vsg::createPhongShaderSet(options);
    if (!phong)
    {
        return;
    }
    phong->defaultShaderHints = shaderHints;
    phong->variants.clear();

    options->shaderSets["phong"] = phong;
```
Models differ only by data, but using same shader, is there any way to create only one GraphicsPipelineConfigurator and will it optimize my program somehow?

**Comment by robertosfield** [comment]
As general rule you want to share as much state in the scene graph as you can, and minimize the number of different GraphicsPipeline that the scene has.  The [vsg::SharedObjects]() class provides support for this, and this is use by the [vsgXchange::assimp](https://github.com/vsg-dev/vsgXchange/blob/master/src/assimp/SceneConverter.cpp) and the [vsg::Builder](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/utils/Builder.cpp) class, this enables sharing of state as as well as the GraphicsPipelineConfigurators.

Another approach is done in the [vsg::tile ReaderWriter](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/io/tile.cpp) which creates a single GraphicsPipelineConfigurator for a tiled databases and then reuses it there after.  This approach is most lightweight but it can get away with as all tiles have the same requirements w.r.t textures and shaders.

**Reply by avilkinnick** [reply]
Is it possible even for pagedLOD?

**Reply by robertosfield** [reply]
;Yes, vsg::tile uses vsg::PagedLOD
